# TP NoSQL

Prénom :

Nom : 

Pour ce TP noté vous avez trois heures pour le réaliser et 48 heures pour le finir si besoin. Le rendu final est attendu pour dimanche 29 septembre 2024 à 23 h 59.

Vous serrez évalués sur les requêtes mais également sur la qualité et la propreté du code. Pour chaque question, une réponse présentant les résultats est attendue.

In [2]:
# installation des packages pandas et pymongo
!pip install -r requirements.txt --quiet

In [3]:
import pymongo
import pandas as pd

URI = 'mongodb+srv://usermongo:jcFu0zqFcwn2prAf@cluster0.57woh.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0'
client = pymongo.MongoClient(URI)
db = client.tp

# output the name of the collections in the database
print("Collections: ", db.list_collection_names())

Collections:  ['movies', 'restaurants']


1. Combien de films sont dans la base de données ?

In [38]:
x = db.movies.count_documents({})
print("il y a",x,"films dans la base de données")

il y a 23530 films dans la base de données


2. Quel est le nombre moyen de votes par pays ? Trier les pays par nombre moyen de votes décroissant.

In [116]:
nombre_moyen_votes_par_pays = db.movies.aggregate([
    {"$unwind":"$countries"}
    {"$group": {"_id": "$countries", "average_votes": {"$avg": "$votes"}}},
    {"$sort": {"average_votes": -1}}
])
pd.DataFrame(list(nombre_moyen_votes_par_pays))

SyntaxError: invalid syntax. Perhaps you forgot a comma? (1485347714.py, line 2)

3. Donner la liste des différents genres de films.

In [48]:
genre_films = db.movies.distinct("genres")
print("voici les différents genre de films",genre_films)

voici les différents genre de films ['Action', 'Adventure', 'Animation', 'Biography', 'Comedy', 'Crime', 'Documentary', 'Drama', 'Family', 'Fantasy', 'Film-Noir', 'History', 'Horror', 'Music', 'Musical', 'Mystery', 'News', 'Romance', 'Sci-Fi', 'Short', 'Sport', 'Talk-Show', 'Thriller', 'War', 'Western']


4. Combien de films ont une note IMDb supérieure à 8 ?

In [52]:
note = db.movies.count_documents({"imdb.rating": {"$gt":8}})
print("il y a", note, "films qui ont une note supérieure a 8")

il y a 1177 films qui ont une note supérieure a 8


5. Afficher le titre, l'année de sortie et les réalisateurs de tous les films (ne pas afficher l'ID).

In [58]:
y = db.movies.find({},{"title":1,"year":1,"directors":1,"_id":0})
pd.DataFrame(list(y))

title  \
0      Intolerance: Love's Struggle Throughout the Ages   
1                             The Poor Little Rich Girl   
2                                               The Kid   
3                                        Miss Lulu Bett   
4                                          Safety Last!   
...                                                 ...   
23525                                            Emelie   
23526                                    Techo y comida   
23527                                             Nahid   
23528                                  Coin Locker Girl   
23529                                 Tryapichnyy soyuz   

                            directors  year  
0                     [D.W. Griffith]  1916  
1                  [Maurice Tourneur]  1917  
2                   [Charles Chaplin]  1921  
3               [William C. de Mille]  1921  
4      [Fred C. Newmeyer, Sam Taylor]  1923  
...                               ...   ...  
23525                [Michael Thelin]  2015  
23526      [Juan Miguel del Castillo]  2015  
23527                [Ida Panahandeh]  2015  
23528                   [Jun-hee Han]  2015  
23529            [Mikhail Mestetskiy]  2015  

[23530 rows x 3 columns]

par exemple le film The Kid a comme réalisateur Charles Chaplin en 1921


6. Afficher le titre et le plot des films réalisés par D.W. Griffith.

In [64]:
z = db.movies.find({"directors":"D.W. Griffith"},{"title":1,"plot":1,"_id":0})
pd.DataFrame(list(z))

plot  \
0  The story of a poor young woman, separated by ...   
1  A greedy tycoon decides, on a whim, to corner ...   
2  A frail waif, abused by her brutal boxer fathe...   
3  The Civil War divides friends and destroys fam...   

                                              title  
0  Intolerance: Love's Struggle Throughout the Ages  
1                                 A Corner in Wheat  
2    Broken Blossoms or The Yellow Man and the Girl  
3                             The Birth of a Nation

il y a eu 4 films réalisés par D.W. Griffin


7. Quels sont les 5 acteurs les plus populaires (ceux qui ont joué dans le plus de films) ?

In [76]:
populaire = db.movies.aggregate([
    {"$unwind": "$cast"},
    {"$group": {"_id": "$cast", "num_movies": {"$sum": 1}}},
    {"$sort": {"num_movies": -1}},
    {"$limit": 5}
])
pd.DataFrame(list(populaire))
    

_id  num_movies
0      Gèrard Depardieu          68
1        Robert De Niro          60
2         Michael Caine          52
3  Marcello Mastroianni          50
4          Bruce Willis          49

gerard depardieu est numéro 1 car il a joué dans 68 films 

8. Quelles sont les 6 années avec le plus de films produits ?

In [82]:
b = db.movies.aggregate([
    {"$group": {"_id": "$year", "num_movies": {"$sum": 1}}},
    {"$sort": {"num_movies": -1}},
    {"$limit": 6}
])
pd.DataFrame(list(b))

_id  num_movies
0  2013        1220
1  2014        1147
2  2012        1109
3  2011        1040
4  2009         990
5  2010         970

2013 est la meilleure année avec 1220 films produits

9. Combien de films ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7 ?

In [92]:
i = db.movies.count_documents({
    "year": {"$lt": 2000},
    "imdb.rating": {"$gt": 7}
})
print("il y a",i,"films qui ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7")

il y a 4747 films qui ont été réalisés avant l'année 2000 et ont une note IMDb supérieure à 7


10. Donner les 20 réalisateurs qui ont réalisé le plus de films (afficher seulement leur nom et le nombre de films).

In [96]:
realisateurs = db.movies.aggregate([
    {"$unwind": "$directors"},
    {"$group": {"_id": "$directors", "num_movies": {"$sum": 1}}},
    {"$sort": {"num_movies": -1}},
    {"$limit": 20}
])
pd.DataFrame(list(realisateurs))

_id  num_movies
0         Woody Allen          40
1           John Ford          35
2         John Huston          34
3       Takashi Miike          34
4       Werner Herzog          33
5     Martin Scorsese          32
6    Alfred Hitchcock          31
7        Sidney Lumet          30
8    Steven Spielberg          29
9     Mario Monicelli          29
10       George Cukor          29
11      Michael Apted          29
12  Steven Soderbergh          28
13          Spike Lee          28
14      Robert Altman          28
15        Wim Wenders          27
16         Johnnie To          27
17    Jean-Luc Godard          27
18          Ken Loach          27
19     Clint Eastwood          27

woody allen a réalisé 40 films et est donc le numéro 1

Woody allen a réalise le plus de film


11. Quelle est la note IMDb moyenne des films sortis après 2010 ?

In [102]:
moyenne = db.movies.aggregate([
    {"$match": {"year": {"$gt": 2010}}},
    {"$group": {"_id": None, "average_rating": {"$avg": "$imdb.rating"}}}
])
pd.DataFrame(list(moyenne))

_id  average_rating
0  None        6.544033

la note IMDB moyenne des films sorti après 2010 est de 6,4

12. (bonus) Afficher les films dont le titre contient le mot "love".

In [104]:
film= db.movies.find({"title": {"$regex": "love", "$options": "i"}}, {"title": 1, "_id": 0})
pd.DataFrame(list(film))

title
0    Intolerance: Love's Struggle Throughout the Ages
1                                      Arise, My Love
2                                   One Night of Love
3                               Love Finds Andy Hardy
4                                     Love Me Tonight
..                                                ...
438                                        Much Loved
439                               Cafè. Waiting. Love
440                    Larry Kramer in Love and Anger
441                                    Our Loved Ones
442                   My Love, Don't Cross That River

[443 rows x 1 columns]

il y en a 443 film qui contiennent le mot Love

13. (bonus) Afficher le titre des films ayant un score Rotten Tomatoes (viewer) supérieur à 75.


In [128]:
superieur = db.movies.find({"tomatoes.rotten": {"$gt": 75}}, {"title": 1, "_id": 0})
pd.DataFrame(list(superieur))


title
0                   Glitter
1            Batman & Robin
2              The Avengers
3       Eye of the Beholder
4            Fantastic Four
..                      ...
712                 Taken 3
713                Blackhat
714  Magic in the Moonlight
715               Insurgent
716             Hot Pursuit

[717 rows x 1 columns]

il y a 717 films ayany un score supérieur a 75